internal/validator/validator.go

```diff
@@ -0,0 +1,302 @@
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get processed epochs of application %v. %w",
+			app.ContractAddress, err,
+		)
+	}
+
+	for _, epoch := range processedEpochs {
+		slog.Info(
+			fmt.Sprintf(
+				"started calculating claim for epoch %d of application %v",
+				epoch.Index, app.ContractAddress,
+			),
+			"service", v,
+		)
+		claim, outputs, err := v.createClaimAndProofs(ctx, epoch)
+		slog.Info(
+			fmt.Sprintf(
+				"finished calculating claim for epoch %d of application %v",
+				epoch.Index, app.ContractAddress,
+			),
+			"service", v,
+		)
+		if err != nil {
+			return err
+		}
+
+		// The Cartesi Machine calculates the root hash of the outputs Merkle
+		// tree after each input. Therefore, the root hash calculated after the
+		// last input in the epoch must match the claim hash calculated by the
+		// Validator. We first retrieve the hash calculated by the
+		// Cartesi Machine...
+		machineClaim, err := v.repository.GetLastInputOutputHash(
+			ctx,
+			epoch.Index,
+			epoch.AppAddress,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to get the machine claim. %w", err)
+		}
+
+		// ...and compare it to the hash calculated by the Validator
+		if *machineClaim != *claim {
+			return errors.New("validator claim does not match Cartesi Machine claim")
+		}
+
+		// update the epoch status and its claim
+		epoch.Status = EpochStatusClaimComputed
+		epoch.ClaimHash = claim
+
+		// store the epoch and proofs in the database
+		if err := v.repository.SetEpochClaimAndInsertProofsTransaction(
+			ctx,
+			epoch,
+			outputs,
+		); err != nil {
+			return fmt.Errorf("failed to store claim and proofs. %w", err)
+		}
+	}
+	return nil
+}
+
+// createClaimAndProofs calculates the claim and proofs for an epoch. It returns
+// the claim and the epoch outputs updated with their hash and proofs. In case
+// the epoch has no outputs, there are no proofs and it returns the pristine
+// claim for the first epoch or the previous epoch claim otherwise.
+func (v *Validator) createClaimAndProofs(
+	ctx context.Context,
+	epoch Epoch,
+) (*Hash, []Output, error) {
+	epochOutputs, err := v.repository.GetOutputsProducedInBlockRange(
+		ctx,
+		epoch.AppAddress,
+		epoch.FirstBlock,
+		epoch.LastBlock,
+	)
+	if err != nil {
+		return nil, nil, fmt.Errorf(
+			"failed to get outputs for epoch %d of application %v. %w",
+			epoch.Index, epoch.AppAddress, err,
+		)
+	}
+
+	previousEpoch, err := v.repository.GetPreviousEpoch(ctx, epoch)
+	if err != nil {
+		return nil, nil, fmt.Errorf(
+			"failed to get previous epoch for epoch %d of application %v. %w",
+			epoch.Index, epoch.AppAddress, err,
+		)
+	}
+
+	// if there are no outputs
+	if len(epochOutputs) == 0 {
+		// and there is no previous epoch
+		if previousEpoch == nil {
+			// this is the first epoch, return the pristine claim
+			claim, _, err := merkle.CreateProofs(nil, MAX_OUTPUT_TREE_HEIGHT)
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to create proofs. %w", err)
```

Comment:
Add `app address` and `epoch index` to message, to make clear for which epoch this originates from

Reply:
Done in [986d2c5](https://github.com/cartesi/rollups-node/pull/554/commits/986d2c5ae59b7067115b9fb27c49873a84ebd1b8)